streamflow-api.yaml

```diff
@@ -0,0 +1,891 @@
+      tags:
+        - Orders
+      parameters:
+        - $ref: '#/components/parameters/userId'
+      responses:
+        200:
+          description: Success
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StreamflowOrderList'
+
+
+  /streamflow/orders/{id}:
+    get:
+      summary: Get existing streamflow order
+      description: |
+        Content of the `links` property depends on the value of the `status` property.
+        If `status /= 'fulfilled'`, the data links are absent.
+      tags:
+        - Orders
+      parameters:
+        - $ref: '#/components/parameters/orderId'
+      responses:
+        200:
+          description: Success
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StreamflowOrderGeneric'
+        404:
+          $ref: '#/components/responses/404NotFound'
+
+
+    delete:
+      summary: Cancel a streamflow order
+      description: |
+        If the streamflow order is not cancellable (not in pending status), response status is 409 Conflict.
+      tags:
+        - Orders
+      parameters:
+        - $ref: '#/components/parameters/orderId'
+      responses:
+        204:
+          $ref: '#/components/responses/204NoContent'
+        401:
+          $ref: '#/components/responses/401Unauthorized'
+        404:
+          $ref: '#/components/responses/404NotFound'
+        409:
+          $ref: '#/components/responses/409Conflict'
+
+  ### Results
+
+  /streamflow/results:
+    get:
+      summary: List currently ready streamflow results
+      tags:
+        - Results
+      responses:
+        200:
+          description: Success
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StreamflowResultList'
+
+  /streamflow/results/{id}:
+    get:
+      summary: Get metadata of a streamflow result
+      tags:
+        - Results
+      parameters:
+        - $ref: '#/components/parameters/resultId'
+      responses:
+        200:
+          description: Success
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StreamflowResult'
+        404:
+          $ref: '#/components/responses/404NotFound'
+
+    delete:
+      summary: Deprecate a streamflow result (mark as no longer needed)
+      tags:
+        - Results
+      parameters:
+        - $ref: '#/components/parameters/resultId'
+      responses:
+        204:
+          $ref: '#/components/responses/204NoContent'
+        401:
+          $ref: '#/components/responses/401Unauthorized'
+        404:
+          $ref: '#/components/responses/404NotFound'
+        409:
+          $ref: '#/components/responses/409Conflict'
+
+
+  /streamflow/results/{id}/annual-cycle/{startYear}-{endYear}:
+    get:
```

Comment:
A different way to handle the annual cycle query that just occurred to me, if one of the REST design goals is to not make the client construct arbitrary URLs: 

- `/streamflow/result/0 `includes a link to `/streamflow/result/0/annual-cycle` (along with other metadata)
- `/streamflow/result/0/annual-cycle` is solely a collection of links to annual cycles for valid climatologies present in this data, so it would have links to, eg:
    - `/streamflow/result/0/annual-cycle/1961-1999`
    - `/streamflow/result/0/annual-cycle/1971-2000`
    - `/streamflow/result/0/annual-cycle/1981-2010`
    - `/streamflow/result/0/annual-cycle/2010-2039`
    - `/streamflow/result/0/annual-cycle/2040-2069`
    - `/streamflow/result/0/annual-cycle/2070-2099`
- Pages like` /streamflow/result/0/annual-cycle/2070-2099` would actually return a timeseries.

Or there could be a climatology resource type, like` /streamflow/result/0/climatology/1960-1989/annual-cycle` or something, but I don't know of any other data analysis we'd like broken out by climatology, so I don't think that buys us anything.

Reply:
You rock. This is an excellent idea.

Reply:
This presumes that one knows in advance what climatologies are of interest. But it can't hurt to have a standard set listed.